sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala

```diff
@@ -22,252 +22,57 @@ import java.io.File
-    unresolved.map(_.show).sorted.foreach { signature =>
-      logger.error(s"cannot link: $signature")
-    }
-
-    throw new MessageOnlyException("unable to link")
-  }
-
-  implicit class RichLogger(logger: Logger) {
-    def time[T](msg: String)(f: => T): T = {
-      import java.lang.System.nanoTime
-      val start = nanoTime()
-      val res   = f
-      val end   = nanoTime()
-      logger.info(s"$msg (${(end - start) / 1000000} ms)")
-      res
-    }
-
-    def running(command: Seq[String]): Unit =
-      logger.debug("running" + nl + command.mkString(nl + "\t"))
-  }
-
-  /** Compiles application nir to llvm ir. */
-  private def compileNir(config: tools.Config,
-                         logger: Logger,
-                         linkerReporter: tools.LinkerReporter,
-                         optimizerReporter: tools.OptimizerReporter,
-                         cwd: File): Seq[nir.Attr.Link] = {
-
-    val driver = tools.OptimizerDriver(config)
-    val (unresolved, links, raw, dyns) = logger.time("Linking") {
-      tools.link(config, driver, linkerReporter)
-    }
-
-    if (unresolved.nonEmpty) { reportLinkingErrors(unresolved, logger) }
-
-    val classCount = raw.count {
-      case _: nir.Defn.Class | _: nir.Defn.Module | _: nir.Defn.Trait => true
-      case _                                                          => false
-    }
-    val methodCount = raw.count(_.isInstanceOf[nir.Defn.Define])
-    logger.info(s"Discovered ${classCount} classes and ${methodCount} methods")
-
-    val optimized = logger.time("Optimizing") {
-      tools.optimize(config, driver, raw, dyns, optimizerReporter)
-    }
-
-    logger.time("Generating intermediate code") {
-      tools.codegen(config, optimized)
-    }
-    logger.info(s"Produced ${(cwd ** "*.ll").get.length} files")
-
-    links
-  }
-
-  /** Compiles *.c[pp] in `cwd`. */
-  def compileCSources(clang: File,
-                      clangpp: File,
-                      cwd: File,
-                      gc: String,
-                      logger: Logger): Boolean = {
+  val nativeTarget =
+    taskKey[String]("Target triple.")
 
-    val cpaths   = filterGCSources(gc, (cwd ** "*.c").get, cwd).map(abs)
-    val cpppaths = filterGCSources(gc, (cwd ** "*.cpp").get, cwd).map(abs)
-    val paths    = cpaths ++ cpppaths
+  val nativeLinkerReporter =
+    settingKey[tools.LinkerReporter](
+      "A reporter that gets notified whenever a linking event happens.")
 
-    paths.par
-      .map { path =>
-        val isCppSource = path.endsWith(".cpp")
+  val nativeOptimizerReporter =
+    settingKey[tools.OptimizerReporter](
+      "A reporter that gets notified whenever an optimizer event happens.")
 
-        val compiler = abs(if (isCppSource) clangpp else clang)
-        val flags    = if (isCppSource) Seq("-std=c++11") else Seq()
+  val nativeOptimizerDriver =
+    taskKey[tools.OptimizerDriver]("Pass manager for the optimizer.")
```

Comment:
An interesting side effect of making everything an sbt task, is that every single part can be overriden. For example one can inject their own NIR passes into the optimizer driver if they really want and are not afraid to use internal APIs.